adapters/metricsReporter.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2016 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package adapters
+
+import "time"
+
+const (
+	// Unspecified is unknown category of metrics.
+	Unspecified MetricKind = 0
+	// Gauge is a metric that represents a single value measurement that can go
+	// up or down arbitrarily.
+	Gauge
+	// Counter is a metric that represents a single value that only ever
+	// increases.
+	Counter
+)
+
+type (
+	// MetricsReporter is the interface for adapters that will handle metrics
+	// data within the mixer.
+	MetricsReporter interface {
+		Adapter
+
+		// ReportMetrics directs a backend adapter to process a batch of
+		// MetricValues derived from potentially several Report() calls.
+		ReportMetrics([]MetricValue) error
+	}
+
+	// MetricKind represents the category of metric that is recorded in a metric
+	// value.
+	MetricKind int
+
+	// MetricValue holds the value of a metric, along with dimensional and
+	// other metadata.
+	MetricValue struct {
```

Comment:
I think we want to separate schema from values, otherwise it is too repetitive. A metric value should say "this value corresponds to schema A", and the details of schema X indicate the description, display name, unit, etc. We can deliver all the descriptors as an map as part of aspect-level config state, and adapters can hold on to the pointer.